### ISSUE TITLE
Backward compatibility with v1.0.0 - artifact extracted to the current working dir and folder name is not preserved.

### ISSUE THREAD

**Issue by jdevelop** [actions/download-artifact#30]
I used the artifacts with my Firebase deployment project, that expects the `dist` folder configured in the `public` section of firebase to contain all the artifacts.

With the following snippet:
```
name: Deploy Firebase
on:
  push:
    branches:
      - master

jobs:
  build:
    name: Build
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repo
        uses: actions/checkout@master
      - name: Install Dependencies
        run: npm install
      - name: Build
        run: npm run build
      - name: Archive Production Artifact
        uses: actions/upload-artifact@v2
        with:
          name: dist
          path: dist
  deploy:
    name: Deploy
    needs: build
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repo
        uses: actions/checkout@master
      - name: Download Artifact
        uses: actions/download-artifact@v2
        with:
          name: dist
      - name: Deploy to Firebase
        uses: w9jds/firebase-action@master
        with:
          args: deploy --only hosting:stage
        env:
          FIREBASE_TOKEN: ${{ secrets.FIREBASE_TOKEN }}
          PROJECT_NAME: name
```

it doesn't work - looks like the `dist` archive either isn't unpacked to the `dist` folder when downloaded. Instead it is unpacked into the current working directory.

Rolling back to `v1.0.0` fixes the problem.

**Comment by konradpabjan** [comment]
Thanks for bringing this up! Surprised no one reported this during the `v2-preview`

Looks like this breaks a few users that are pinned to `@master` 😥 (never ideal) but I somewhat warned about the possibility of this: https://github.com/actions/download-artifact/issues/15

I'm working on a quick new option that will allow users to specify in the YAML if they want an extra folder.

**Comment by tedmiston** [comment]
This one broke my pipelines as well.  I like to keep pinned to master (for a side project).  Rolling back to v1 fixed it for now.

**Comment by chancancode** [comment]
This certainly broke things, but given it's a major version bump, that is probably acceptable. I'm not sure that this should be "fixed", though. Since there is already a `path` option, you can always get the previous behavior by specifying `path: (name of artifact)`, which seems easy enough. 

However, the reverse is not true, if the folder is added and you don't want it, you can't fix that by changing the path, and you will need another flag (similar to the `--strip-components` flag in `tar`) to control that which is pretty awkward and unnecessary in my opinion.

Not adding a folder also matches the "convention" of a few other official actions, including checkout, which checks out the repository at `$GITHUB_WORKSPACE` by default and allowing to customize that with `path` (which is relative to the workspace). That is essentially the semantics here in v2.

Even though this change broke my workflows (I pinned to `@master`, so that's probably my fault), I strongly prefer to keep this new behavior.

This is essentially the flip side of https://github.com/actions/upload-artifact/issues/30, which is somewhat relevant to this.

**Comment by chancancode** [comment]
Put it another way, here are the possible options:

## Option 1: Never wrap (the current v2 behavior)

If you don't want the folder, you do this:

```yaml
- uses: actions/download-artifact@v2
  with:
    name: dist
```

If you want the folder, you do this:

```yaml
- uses: actions/download-artifact@v2
  with:
    name: dist
    path: dist
```

*Seems good to me.*

## Option 2: Always wrap (the v1 behavior)

If you don't want the folder, you are out of luck, as far as this action is concerned. You can workaround it by running a subsequent command that moves everything up one folder, or specify a working directory for every subsequent steps.

If you want the folder, you do this:

```yaml
- uses: actions/download-artifact@v2
  with:
    name: dist
```

*Seems bad, even though we happen to be used to it.*

## Option 3: Make it optional, defaulting to no wrap

If you don't want the folder, you do this:

```yaml
- uses: actions/download-artifact@v2
  with:
    name: dist
```

If you want the folder, you do this:

```yaml
- uses: actions/download-artifact@v2
  with:
    name: dist
    wrap: true
```

...or this...

```yaml
- uses: actions/download-artifact@v2
  with:
    name: dist
    path: dist
```

*Seems silly, since there are now two ways to accomplish the same thing, more documentation to read and write. If you are in the second camp, you are still initially surprised it doesn't do what you want, have to go read the documentation and contemplate between the two equivalent options, and it's not like the `warp: true` route is saving you from a ton of typing.*

## Option 4: Make it optional, defaulting to wrap

If you don't want the folder, you do this:

```yaml
- uses: actions/download-artifact@v2
  with:
    name: dist
    wrap: false
```

If you want the folder, you do this:

```yaml
- uses: actions/download-artifact@v2
  with:
    name: dist
```

*This is... okay, I guess. Personally I would argue this seems unnecessary, introduces another thing to document and learn, when the existing option is already expressive enough for all use cases. I would also that it is surprising and unexpected, since manually downloading the zip file doesn't include the wrapper folder anymore since https://github.com/actions/upload-artifact/issues/30, and other actions like checkout doesn't automatically add a wrapping folder also. However, at least it allows me to express what I want, so it's not the end of the world.*

**Comment by konradpabjan** [comment]
Thanks for the suggestion @chancancode 

We're going to do something like option 3 (for `v2` keep it so there is no folder by default, and add an extra option so that a folder can be created if needed). This will mean a permanent compat difference between `v1` and `v2`, but at this point there is no going back. If users do want a folder, they will be able to specify `true`.

Another bonus of having an extra option in the YAML is that it will give users flexibility when we add support for things like downloading multiple artifacts via a wildcard pattern. Right now there is an option to download all artifacts at once, however a separate folder gets created for each artifact. I've seen a few users already moving all the artifact contents into a single folder after everything is downloaded so an extra option in the YAML to specify the behavior would be useful for that scenario as well.

**Comment by konradpabjan** [comment]
I've added a section to the `README` about the compat difference between `v1` and `v2`.
https://github.com/actions/download-artifact#compatibility-between-v1-and-v2

For anyone that was using `@master` and was broken, it should be very simple to get the same behavior as before (set the `path` input to the name of the `artifact` that was being downloaded).

Since v2 is a new release, we are okay with the compat difference and we won't be changing it to how it was before.

**Comment by chancancode** [comment]
@konradpabjan Thanks for considering it! FWIW, I am affected by it but I am happy with the change! 😄 

> Thanks for bringing this up! Surprised no one reported this during the v2-preview

For future, it may help if this was developed on master, or at least have landed on master for a while before v2 was released, so users who pinned to master can help test these changes and report issues. Intentionally or otherwise, that's kind of what we signed up for.

(Though to be fair, I _think_ the reason this was so widespread was because when this was first made available, there was no v1 release yet, and the README at the time suggested using master, IIRC.)

Anyway, this really isn't a big deal (for me), and thanks again 😄

**Comment by konradpabjan** [comment]
So behind the scenes there were a number of factors that made this `v2` transition "rough" (it should be orders of magnitude smoother moving forward).

If you look at the `v1` version of this action, there are 3 files in this [repo](https://github.com/actions/download-artifact/tree/releases/v1) 🙃 The `v1` version of this action is "special" because it is effectively a C# plugin that is coupled to the actions runner.

Normal actions are all typescript/javascript, but the `v1` version was originally a plugin because the runner had some special permissions and APIs to interact with artifacts to make the whole upload/download thing work. 

To make `v2` possible, we had to do a whole bunch of API work behind the scenes to allow normal typescript/javascript actions to interact with artifacts. Then we had to actually port over all the upload/download logic to typescript (it's quite complex, especially on the upload side).

If we developed this all on `@master`, we would have instantly broken a lot of users as soon as development started who pinned to `@master` instead of `@v1`. Our telemetry was telling us that a significant percentage of download (and upload) artifact users were pinned to `@master` so we went down the route of a `@v2-preview` on a separate branch so we wouldn't break so many people early on.

Our docs and README early on recommended `@master` and we didn't change that until a significant amount of users had already started using the actions 😕 We also don't have any feature or easy way of letting users know that they should NOT be using a specific version (this includes deprecation). This is where I think we had a problem. Most users didn't break because they started using `@v2`, it's because as soon as we merged into master, anyone pinned to `@master` got the new version. 

This compat issue thankfully only effected users who didn't use a `path` input in `v1` so it could have been worse
